File: backend/saas/legacyCrmGuard.ts
import { SAAS_CRM_SOURCE_OF_TRUTH } from "./crmConsolidation";

const WARNED = new Set<string>();

/**
 * Fase 1B — aviso en runtime para escrituras CRM legacy.
 * Nuevas features deben usar SaasCrmService / saas_contacts.
 */
export function warnLegacyCrmWrite(caller: string, table: "contacts" | "crm_contacts"): void {
  const key = `${caller}:${table}`;
  if (WARNED.has(key)) return;
  WARNED.add(key);
  const msg =
    `[NELVYON CRM LEGACY] ${caller} escribe en "${table}". ` +
    `Use ${SAAS_CRM_SOURCE_OF_TRUTH.service} (${SAAS_CRM_SOURCE_OF_TRUTH.table}) para desarrollo nuevo. ` +
    `Ver docs/PHASE_1B_CRM_ETL.md`;
  if (typeof process !== "undefined" && process.env?.NODE_ENV !== "test") {
    console.warn(msg);
  }
}

export function resetLegacyCrmWriteWarningsForTests(): void {
  WARNED.clear();
}

File: backend/services/legacy_crm_guard.py
"""Fase 1B — guardrails para escrituras CRM legacy (Python)."""

from __future__ import annotations

import logging
import warnings

logger = logging.getLogger(__name__)

_WARNED: set[str] = set()

SAAS_CRM_OFFICIAL = "saas_contacts / SaasCrmService"


def warn_legacy_crm_write(caller: str, table: str) -> None:
    key = f"{caller}:{table}"
    if key in _WARNED:
        return
    _WARNED.add(key)
    msg = (
        f"[NELVYON CRM LEGACY] {caller} escribe en '{table}'. "
        f"Use {SAAS_CRM_OFFICIAL} para desarrollo nuevo. "
        "Ver docs/PHASE_1B_CRM_ETL.md"
    )
    warnings.warn(msg, stacklevel=3, category=DeprecationWarning)
    logger.debug(msg)
